### DISCUSSION TITLE
Refactor L2CAP API?

### DISCUSSION THREAD

**Discussion by zxzxwu** [google/bumble#246] [General]
The current L2CAP APIs have some issues, like in `device`
* `create_l2cap_connector()`
* `open_l2cap_channel()`
* `register_l2cap_server()`
* `register_l2cap_channel_server()`

These methods are quite unreadable. Developers cannot understand which transport they indicate unless tracing the code.

The same problem also happens inside `l2cap` module:
* The class name doesn't match their actual presence in the spec(`Channel`->Classic Channel, `LeConnectionOrientedChannel`-> LE Credit Based Channel)
* Classes don't inherit a common interface, which may increase a diversity
* Missing methods to deregister dynamic channels
* Unknown typing of PDU(this requires refactor in `host`<->`device`<->`l2cap`)

Thus, as we are receiving some requests like Pandora, EATT, and Enhanced Credit Based Channel, in this thread I would like to have your opinion whether and how we should refactor the L2CAP functionality in Bumble.

**Comment by zxzxwu** [comment]
@uael @DeltaEvo @barbibulle

**Reply by DeltaEvo** [reply]
I'm not maintaining Bumble so my opinion don't matter much here, but seeing your arguments I agree that it would be good to change that

**Comment by barbibulle** [comment]
Yes, definitely worth discussion improvements and changes. I'll set up some time to talk.
